core/src/main/java/contacts/core/util/WhereSubstitutions.kt

```diff
@@ -0,0 +1,269 @@
+        Where(
+            lhs = substituteField(lhs),
+            operator = operator,
+            rhs = substituteValue(rhs),
+            options = options,
+            isRedacted = isRedacted
+        )
+    } else if (lhs is WhereHolder && operator is Operator.Combine && rhs is WhereHolder) {
+        // Recursive case. Traverse tree.
+        Where(
+            lhs = WhereHolder(
+                lhs.where.copyWithFieldValueSubstitutions(
+                    substituteField,
+                    substituteValue
+                )
+            ),
+            operator = operator,
+            rhs = WhereHolder(
+                rhs.where.copyWithFieldValueSubstitutions(
+                    substituteField,
+                    substituteValue
+                )
+            ),
+            options = options,
+            isRedacted = isRedacted
+        )
+    } else {
+        // This should not happen with the current structure. If this does happen, it means
+        // that we made some changes that broke the structure.
+        throw ContactsException(
+            "Unhandled Where form lhs: ${lhs.javaClass.simpleName}, rhs: ${rhs.javaClass.simpleName}"
+        )
+    }
+
+/**
+ * Returns a "reduced" copy of the given [where] clause to ensure that it is able match rows in the
+ * Data table.
+ *
+ * The following will always match 0 rows in the Data table,
+ *
+ * ```kotlin
+ * Email.Address.isNotNull() and Phone.Number.isNotNull()
+ * ```
+ *
+ * because it is impossible to have a row in the Data table that represents both an email and a
+ * phone number! Each row in the data table can only have one mimetype!
+ *
+ * > Note that it does not matter what operator is used (it does not have to be "isNotNull").
+ *
+ * Therefore, we need to perform a Data table query for each statement in the where clause that;
+ *
+ * - are combined using "AND"
+ * - when the mime types are different
+ *
+ * Example #1,
+ *
+ * where  -> Email and Phone
+ * reduce -> ContactIDs and Phone
+ *
+ * Example #2,
+ *
+ * where  -> (((Email and Phone) and Organization) and Website)
+ * reduce  -> (((ContactIDs and Phone) and Organization) and Website)
+ * reduce  -> ((ContactIDs and Organization) and Website)
+ * reduce  -> (ContactIDs and Website)
+ *
+ * Example #3,
+ *
+ * where  -> (Email and Phone) and (Organization and Website)
+ * reduce -> (ContactIDs and Phone) and (Organization and Website)
+ * reduce -> ContactIDs and (Organization and Website)
+ * reduce -> ContactIDs and (ContactIDs and Website)
+ *
+ * Example #4,
+ *
+ * where  -> (Email or Phone) and (Organization or Website)
+ * reduce -> (ContactIDs) and (Organization or Website)
+ *
+ * The idea is to reduce the amount of mimetype-aware matches such that there is 0 mimetype-aware
+ * match in one side of the AND.
+ *
+ * More context in https://github.com/vestrel00/contacts-android/issues/142#issuecomment-1000948529
+ */
+internal fun ContentResolver.reduce(
```

Comment:
This is where the magic happens. I'll let the documentation and code speak for themselves ❤️